kafkaesk/ext/logging/formatter.py

```diff
@@ -0,0 +1,81 @@
+from .record import factory
+from .record import PydanticLogRecord
+from typing import Any
+from typing import Dict
+from typing import Optional
+from typing import Tuple
+from typing import Type
+
+import logging
+import pydantic
+
+
+class BaseLogFormat(pydantic.BaseModel):
+    timestamp: str = pydantic.Field(alias="asctime")
+    logger: str = pydantic.Field(alias="name")
+    severity: str = pydantic.Field(alias="levelname")
+    level: int = pydantic.Field(alias="levelno")
+    message: str
+    exception: Optional[str] = pydantic.Field(alias="exc_type")
+    trace: Optional[str] = pydantic.Field(alias="exc_text")
+    stack: Optional[str] = pydantic.Field(alias="stack_info")
+
+
+class PydanticLogModel(pydantic.BaseModel):
+    class Config:
+        extra = pydantic.Extra.allow
+
+
+class PydanticFormatter(logging.Formatter):
+    def __init__(
+        self,
+        model: Type[pydantic.BaseModel] = BaseLogFormat,
+        fmt: Optional[str] = None,
+        datefmt: Optional[str] = None,
+        style: str = "%",
+    ):
+        self.format_class = model
+        self._init_logrecord_factory()
+        super().__init__(fmt=fmt, datefmt=datefmt, style=style)
+
+    def _init_logrecord_factory(self) -> None:
+        if logging.getLogRecordFactory() != factory:
+            logging.setLogRecordFactory(factory)
+
+    def _format_base_log(self, record: PydanticLogRecord) -> Dict[str, Any]:
+        return self.format_class(**record.__dict__).dict(exclude_none=True)
+
+    def _format_extra_logs(self, record: PydanticLogRecord) -> Dict[str, Any]:
+        extra_logs: Dict[str, Any] = {}
+        if getattr(record, "_pydantic_data", None) is None:
```

Comment:
What's the motivation for accessing an internal attribute and using the `gettattr` for doing that?, looking into the signature seems that the input `record` will be always a `PydanticLogRecord`, why not just do something like?

```python
record._pydantic_data
```

Or just have a public method for accessing to the extra data.

Reply:
Again, good call here.  Initially `_format_extra_logs` was doing more work that eventually got moved to the `factory` function and those checks can be removed.